File: Java/Array & Hashing/ValidAnagram.java
/*
 * Given two strings s and t, return true if t is an anagram of s, and false otherwise.

An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

 

Example 1:

Input: s = "anagram", t = "nagaram"
Output: true
Example 2:

Input: s = "rat", t = "car"
Output: false
 

Constraints:

1 <= s.length, t.length <= 5 * 104
s and t consist of lowercase English letters.
 

Follow up: What if the inputs contain Unicode characters? How would you adapt your solution to such a case?
 */

import java.util.HashMap;
import java.lang.*;
public class ValidAnagram {

    public static void main(String args[])
    {
        String s = "anagram";
        String t = "nagaram";
        //s= "rat"; t ="cat";
        boolean value = isAnagram(s,t);
        System.out.println(value);
    }

    public static boolean isAnagram(String s, String t)
    {
        HashMap<Character,Integer> map = new HashMap<Character,Integer>();
        int len1 = s.length();
        int len2 = t.length();
        if(len1 != len2)
            return false;

        for(int i = 0; i< len1; i++)
        {
            char c = s.charAt(i);
            if(map.containsKey(c)){
                map.put(c,map.get(c)+1);
            } else {
                map.put(c,0);
            }
        }

        for(int j =0; j<len2; j++)
        {
            char d = t.charAt(j);
            if(map.containsKey(d)){
                map.put(d,map.get(d)-1);
                if(map.get(d) < 0 ) {
                    map.remove(d);
                }
            } else {
                return false;
            }
        }

        // check the map should be empty 
        if(map.isEmpty())
            return true;
        else 
            return false;

    }


    
}

File: Kotlin/ValidAnagram.kt
class ValidAnagram {


    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val  s: String = "anagram";
            val  t: String = "nagaram";
        
            val result : Boolean  = isAnagram(s,t);
            println(result);
        }
        fun isAnagram(s: String, t: String): Boolean {
            val len1 = s.length;
            val len2 = t.length;
            val map = mutableMapOf<Char,Int?>();
            if(len1 != len2){
                print("why?")
                return false
            }
            
                val sc: CharArray = s.toCharArray()
                val tc: CharArray = t.toCharArray()
            for(i in sc.indices)
            {
                val c = sc.get(i)
                if(map.containsKey(c)){
                    var value : Int? = map.get(c);
                    map.put(c,value?.let { value+1 });
                    println("CCCC adding +1 ${map.get(c)}")
                } else {
                    map.put(c,0);
                    println("CCCC when you put $c ${map.get(c)}")
                }
            }
            for(j in tc.indices)
            {
                val d = tc.get(j)
                if(map.containsKey(d)){
                    // Decrease the count to 1 and if count is 0 then remove it from map 
                    var value : Int? = map.get(d);
                    map.put(d, value?.let { value - 1 })
                    var after = map.get(d);
                    println("DDDD when you sub - 1 ${map.get(d)}")
                    if(after != null && after < 0 )
                    {
                        map.remove(d)
                        println("DDDD remove  ${map.get(d)}")
                    }
                } else {
                    println("DDDD not in map ")
                    return false
                }
            }
            // check the map should be empty 
            if(map.isEmpty())
                return true;
            else 
                return false;
            
        }
    }
}